t/lib/Test/Act/Environment.pm

```diff
@@ -0,0 +1,221 @@
+# ABSTRACT: Create an environment for testing the Act application
+use 5.20.0;
+package Test::Act::Environment;
+
+use Moo;
+use Types::Standard qw(InstanceOf Str);
+use namespace::clean;
+
+use feature qw(signatures);
+no warnings qw(experimental::signatures);
+
+use File::Temp qw(tempdir);
+use File::Copy::Recursive qw(dircopy);
+
+use FindBin qw($RealBin);
+FindBin::again();
+
+use Test::WWW::Mechanize::PSGI;
+
+use Test::Lib;
+use Test::Act::SMTP::Server;
+
+my $tempdir;
+
+# This needs to happen _before_ Act::Config is used...
+BEGIN {
+    $tempdir = File::Temp->newdir( CLEANUP => 0 );
+    # The following will need adaption to the post-november master
+    # Part 1: Link the distribution files from the repository so that
+    # changes in the working tree become immediately effective
+    for my $dir (qw(templates po wwwdocs)) {
+        symlink "$RealBin/../$dir","$tempdir/$dir"
+            or die "Failed to create a symlink for '$RealBin/../$dir': '$!'";
+    }
+    # Part 2: Copy the files from the test environment
+    for my $dir (qw(actdocs conf conferences)) {
+        dircopy "$RealBin/acthome/$dir","$tempdir/$dir";
+    }
+    # Part 3: These must just exist
+    for my $dir (qw(photos ttc)) {
+        mkdir "$tempdir/$dir" or die "Could not create '$tempdir/$dir': '$!'";
+    }
+    $ENV{ACTHOME} = "$tempdir";
+}
+
+use Act::Config;
+
+my $smtp_server = Test::Act::SMTP::Server->instance;
+$Config->set(email_hostname => 'localhost');
+$Config->set(email_port     => $smtp_server->port);
+
+
+has mech => (
+    is => 'rwp', isa => InstanceOf['Test::WWW::Mechanize'],
+    builder => 'build_mech',
+    documentation =>
+        'A test client for the application tests',
+);
+
+sub build_mech ($self) {
+    require Act::Dispatcher;
+    my $mech = Test::WWW::Mechanize::PSGI->new(app => Act::Dispatcher->to_app);
+    $self->_set_mech($mech);
```

Comment:
This is a kind of a funny pattern. You set the buider of `mech` to `build_mech` and then say `_set_mech` within the builder. So people calling `build_mech` now also set mech, which you set to `rwp`, which disables the public setter, which sort of is `build_mech` now.  Personally I would set `mech` to `ro`, set the builder to `1` (which then makes the builder `_build_mech` although I'm also ok with the current builder, if you want to have it public) and not set `_set_mech($mech)` but just let the builder return `$mech`.

Reply:
Good point, that needs to be fixed.  I used to have a sort-of constant mech created by the builder when I realized that I need to be able to renew a mech.  I intentionally have no public setter, though, because this would need the module's user to create the mech.